pkg/kine/drivers/generic/generic.go

```diff
@@ -381,6 +395,142 @@ func (d *Generic) queryRowPrepared(ctx context.Context, txName, sql string, prep
 	return r
 }
 
+func (d *Generic) InitializeWriteBatching(ctx context.Context) {
+
+	if !d.BatchingEnabled {
+		return
+	}
+
+	if d.BatchingInterval < 10*time.Millisecond {
+		logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Interval is less than 10ms, disabling batching of insert queries")
+		return
+	}
+
+	if d.BatchingMaxQueries < 5 {
+		logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Queue size is less than 5, disabling batching of insert queries")
+		return
+	}
+
+	logrus.Warn("Batching of write queries is currently experimental and MUST NOT be used in production. Expect data loss.")
+	logrus.WithFields(logrus.Fields{"interval": d.BatchingInterval, "queue_size": d.BatchingMaxQueries}).Print("Enabling batching of insert queries")
+	d.flushCh = make(chan struct{})
+	d.addToBatchQueueCh = make(chan BatchedInsert)
+	d.removeFromBatchQueueCh = make(chan int)
+	d.batchingQueue = make([]BatchedInsert, 0, d.BatchingMaxQueries)
+	d.batchingCond = sync.Cond{L: &d.batchingMu}
+
+	go d.watchBatchQueue(ctx)
+	go d.batchQueueWatcher(ctx)
+
+}
+
+func (d *Generic) batchQueueWatcher(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case item := <-d.addToBatchQueueCh:
+			d.batchingCond.L.Lock()
+			if len(d.batchingQueue) >= d.BatchingMaxQueries {
+				d.flushCh <- struct{}{}
+				d.batchingCond.Wait()
+			}
+			d.batchingQueue = append(d.batchingQueue, item)
```

Comment:
The unlucky query that will trigger the flushing will have to wait two flush cycles, one to empty the queue and a second to flush the queue it was placed to. What if we first append the query and then see if we need to flush?